### PR TITLE
simplify training phase as Enum

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

simplify the training phase logic
application in code will follow

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [x] **Check that target branch and milestone match!**

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/states.py [hunk 1]
```diff
@@ -17,9 +17,10 @@
 from typing import Callable, Optional
 
 import pytorch_lightning
+from pytorch_lightning.utilities import LightningEnum
 
 
-class TrainerState(str, Enum):
+class TrainerState(LightningEnum):
```
Comment (carmocca) [inline pytorch_lightning/trainer/states.py:23]: In the docs, it says:

```
    >>> # which is case sensitive
    >>> TrainerState.FINISHED == 'finished'
    False
```

which is wrong with `LightningEnum`. Can you update it?

File: pytorch_lightning/trainer/states.py [hunk 2]
```diff
@@ -36,6 +37,17 @@ class TrainerState(str, Enum):
     INTERRUPTED = 'INTERRUPTED'
 
 
+class TrainPhase(LightningEnum):
```
Comment (carmocca) [inline pytorch_lightning/trainer/states.py:40]: I don't like the name. `testing` is not a training phase...

Call it `RunningStage`?

File: pytorch_lightning/trainer/states.py [hunk 3]
```diff
@@ -36,6 +37,17 @@ class TrainerState(str, Enum):
...
+    """Type of train phase.
+
+    >>> # you can math the type with string
```
Comment (carmocca) [inline pytorch_lightning/trainer/states.py:43]:
```suggestion
    >>> # you can match the Enum with string
```

**Comment by carmocca** [comment]
Can we unify this with LoggerStages?

File: pytorch_lightning/trainer/deprecated_api.py [hunk 4]
```diff
@@ -11,14 +11,15 @@
...
-
+from pytorch_lightning.trainer.states import RunningStage
 from pytorch_lightning.utilities import DistributedType, DeviceType, rank_zero_warn
 
 
 class DeprecatedDistDeviceAttributes:
 
     _distrib_type: DistributedType
     _device_type: DeviceType
+    _train_phase: RunningStage
```
Comment (carmocca) [inline pytorch_lightning/trainer/deprecated_api.py:22]:
```suggestion
    _train_stage: RunningStage
```

File: pytorch_lightning/trainer/states.py [hunk 5]
```diff
@@ -28,14 +29,25 @@ class TrainerState(str, Enum):
     True
     >>> # which is case sensitive
```
Comment (carmocca) [inline pytorch_lightning/trainer/states.py:30]:
```suggestion
    >>> # which is case insensitive
```

File: pytorch_lightning/trainer/states.py [hunk 6]
```diff
@@ -28,14 +29,25 @@ class TrainerState(str, Enum):
...
-    False
+    True
     """
     INITIALIZING = 'INITIALIZING'
     RUNNING = 'RUNNING'
     FINISHED = 'FINISHED'
     INTERRUPTED = 'INTERRUPTED'
 
 
+class RunningStage(LightningEnum):
```
Comment (carmocca) [inline pytorch_lightning/trainer/states.py:40]: Can you refactor and remove

https://github.com/PyTorchLightning/pytorch-lightning/blob/d583d561690ea975ebd9ea41e1b7054495aa0bcb/pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py#L23-L41

to use this class?

**Review by carmocca — APPROVED** [review]

**Review by SkafteNicki — APPROVED** [review]

**Review by s-rog — APPROVED** [review]

File: pytorch_lightning/trainer/states.py [hunk 7]
```diff
@@ -12,30 +12,41 @@
...
+    """Type of train phase.
+
+    >>> # you can match the Enum with string
+    >>> RunningStage.TRAINING == 'train'
+    True
+    """
+    TRAINING = 'train'
+    TESTING = 'test'
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/trainer/states.py:47]: And add them to properties.

**Review by tchaton — COMMENTED** [review]
Great addition ! Let's use this to refactor LoggerConnector internal Stage.
Also, we should provided context manager or stage stack as we can run evaluation in training.

**Review by justusschock — APPROVED** [review]

**Comment by Borda** [comment]
> Great addition ! Let's use this to refactor LoggerConnector internal Stage.
> Also, we should provided context manager or stage stack as we can run evaluation in training.

yes, that is my goal, but as it is widely requested we shall do these changes in smaller steps :]

File: pytorch_lightning/trainer/states.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Borda) [reply pytorch_lightning/trainer/states.py:47]: ok, lest add these stages now but properties when they are needed, so in next PRs

**Comment by Borda** [comment]
@tchaton mind recheck...

**Review by tchaton — APPROVED** [review]

**Pull request merged**